File: apps/ios/Sources/OlloCrypto/IdentityStore.swift
import Foundation

/// Encrypted identity / session / outbox / prekey blobs. The wrapping key never
/// lives next to the ciphertext: on device it comes from Keychain;
/// tests inject a software key.
///
/// Optional `directory` persists already-wrapped files (one per key).
/// Wipe shreds memory and files.
public final class IdentityStore: @unchecked Sendable {
    private var kv: [String: Data] = [:]
    private let directory: URL?

    public init(directory: URL? = nil) {
        self.directory = directory
        if let directory {
            try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
            if let files = try? FileManager.default.contentsOfDirectory(
                at: directory,
                includingPropertiesForKeys: nil
            ) {
                for f in files where !f.lastPathComponent.hasSuffix(".tmp") {
                    if let data = try? Data(contentsOf: f) {
                        kv[f.lastPathComponent] = data
                    }
                }
            }
        }
    }

    public func put(wrapKey: Data, slot: Slot, plaintext: Data) throws {
        try put(wrapKey: wrapKey, key: slot.rawValue, plaintext: plaintext)
    }

    public func put(wrapKey: Data, key: String, plaintext: Data) throws {
        try assertSafeStoreKey(key)
        let sealed = try AesGcmWrap.seal(key: wrapKey, plaintext: plaintext)
        kv[key] = sealed
        try writeFile(key: key, sealed: sealed)
    }

    public func get(wrapKey: Data, slot: Slot) throws -> Data? {
        try get(wrapKey: wrapKey, key: slot.rawValue)
    }

    public func get(wrapKey: Data, key: String) throws -> Data? {
        try assertSafeStoreKey(key)
        guard let blob = kv[key] else { return nil }
        return try AesGcmWrap.open(key: wrapKey, blob: blob)
    }

    public func remove(key: String) throws {
        try assertSafeStoreKey(key)
        kv[key]?.resetBytes(in: 0..<(kv[key]?.count ?? 0))
        kv.removeValue(forKey: key)
        if let directory {
            try? FileManager.default.removeItem(at: directory.appendingPathComponent(key))
        }
    }

    public func keys(prefix: String = "") -> [String] {
        kv.keys.filter { $0.hasPrefix(prefix) }.sorted()
    }

    @discardableResult
    public func wipe() -> Int {
        let n = kv.count
        for k in kv.keys {
            kv[k]?.resetBytes(in: 0..<(kv[k]?.count ?? 0))
        }
        kv.removeAll()
        if let directory, let files = try? FileManager.default.contentsOfDirectory(
            at: directory,
            includingPropertiesForKeys: nil
        ) {
            for f in files {
                if let data = try? Data(contentsOf: f), !data.isEmpty {
                    let shred = Data(count: min(data.count, 4096))
                    try? shred.write(to: f, options: .atomic)
                }
                try? FileManager.default.removeItem(at: f)
            }
        }
        return n
    }

    public var isEmpty: Bool { kv.isEmpty }

    public enum Slot: String, Sendable {
        case identity = "identity.v1"
        case sessions = "sessions.v1"
        case outbox = "outbox.v1"
        case senderKeys = "senderkeys.v1"
        case knownIdentities = "known.v1"
        case preKeys = "prekeys.v1"
        case signedPreKeys = "signedprekeys.v1"
        case threads = "threads.v1"
        case replay = "replay.v1"
        case account = "account.v1"
    }

    private func writeFile(key: String, sealed: Data) throws {
        guard let directory else { return }
        let target = directory.appendingPathComponent(key)
        let tmp = directory.appendingPathComponent("\(key).tmp")
        try sealed.write(to: tmp, options: .atomic)
        if FileManager.default.fileExists(atPath: target.path) {
            try FileManager.default.removeItem(at: target)
        }
        try FileManager.default.moveItem(at: tmp, to: target)
    }
}

func assertSafeStoreKey(_ key: String) throws {
    if key.isEmpty { throw StoreKeyError.empty }
    if key.contains("/") || key.contains("\\") || key.contains("\0") || key.contains("..") {
        throw StoreKeyError.unsafe
    }
}

enum StoreKeyError: Error {
    case empty
    case unsafe
}

File: apps/android/core/crypto/src/main/java/app/ollo/crypto/IdentityStore.kt
package app.ollo.crypto

import java.io.File

/**
 * Encrypted identity / session / outbox / prekey blobs. The wrapping key never
 * lives next to the ciphertext: on device it comes from Android Keystore;
 * tests inject a software key.
 *
 * Optional [directory] persists already-wrapped files (one per key) so a
 * process restart does not drop ratchet state. Wipe shreds memory and files.
 */
class IdentityStore(
    private val kv: MutableMap<String, ByteArray> = linkedMapOf(),
    private val directory: File? = null,
) {
    init {
        directory?.mkdirs()
        directory?.listFiles()?.forEach { f ->
            if (f.isFile && !f.name.endsWith(".tmp") && !kv.containsKey(f.name)) {
                kv[f.name] = f.readBytes()
            }
        }
    }

    fun put(wrapKey: ByteArray, slot: Slot, plaintext: ByteArray) {
        put(wrapKey, slot.key, plaintext)
    }

    fun put(wrapKey: ByteArray, key: String, plaintext: ByteArray) {
        assertSafeStoreKey(key)
        val sealed = AesGcmWrap.seal(wrapKey, plaintext)
        kv[key] = sealed
        writeFile(key, sealed)
    }

    fun get(wrapKey: ByteArray, slot: Slot): ByteArray? = get(wrapKey, slot.key)

    fun get(wrapKey: ByteArray, key: String): ByteArray? {
        assertSafeStoreKey(key)
        val blob = kv[key] ?: return null
        return AesGcmWrap.open(wrapKey, blob)
    }

    fun remove(key: String) {
        assertSafeStoreKey(key)
        kv[key]?.fill(0)
        kv.remove(key)
        directory?.let { File(it, key).delete() }
    }

    fun keys(prefix: String = ""): List<String> = kv.keys.filter { it.startsWith(prefix) }.sorted()

    fun wipe(): Int {
        val n = kv.size
        kv.keys.toList().forEach { k ->
            kv[k]?.fill(0)
            kv.remove(k)
        }
        kv.clear()
        directory?.listFiles()?.forEach { f ->
            if (f.isFile) {
                val nBytes = f.length().toInt().coerceAtMost(4096)
                if (nBytes > 0) f.writeBytes(ByteArray(nBytes))
                f.delete()
            }
        }
        return n
    }

    fun isEmpty(): Boolean = kv.isEmpty()

    private fun writeFile(key: String, sealed: ByteArray) {
        val dir = directory ?: return
        val target = File(dir, key)
        val tmp = File(dir, "$key.tmp")
        tmp.writeBytes(sealed)
        if (target.exists()) target.delete()
        if (!tmp.renameTo(target)) {
            target.writeBytes(sealed)
            tmp.delete()
        }
    }

    enum class Slot(val key: String) {
        Identity("identity.v1"),
        Sessions("sessions.v1"),
        Outbox("outbox.v1"),
        SenderKeys("senderkeys.v1"),
        KnownIdentities("known.v1"),
        PreKeys("prekeys.v1"),
        SignedPreKeys("signedprekeys.v1"),
        Threads("threads.v1"),
    }
}

internal fun assertSafeStoreKey(key: String) {
    require(key.isNotEmpty()) { "empty store key" }
    require(key.none { it == '/' || it == '\\' || it == '\u0000' }) { "unsafe store key" }
    require(!key.contains("..")) { "unsafe store key" }
}
